cutplace/ranges.py

```diff
@@ -442,3 +449,246 @@ def validate(self, name, value, location=None):
+                        raise errors.InterfaceError(
+                            "lower range %d must be greater or equal than upper range %d" % (lower, upper))
+                    result = (lower, upper)
+                else:
+                    # Handle "x".
+                    result = (lower, lower)
+                if result is not None:
+                    for item in self._items:
+                        if self._items_overlap(item, result):
+                            # TODO: use _repr_item() or something to display item in error message.
+                            raise errors.InterfaceError(
+                                "range items must not overlap: %r and %r"
+                                % (self._repr_item(item), self._repr_item(result)))
+                    self._items.append(result)
+                if _tools.is_eof_token(next_token):
+                    end_reached = True
+
+            self._lower_limit = None
+            self._upper_limit = None
+
+            is_first_item = True
+
+            for lower_item, upper_item in self._items:
+                if is_first_item:
+                    self._lower_limit = lower_item
+                    self._upper_limit = upper_item
+                    is_first_item = False
+
+                if lower_item is None:
+                    self._lower_limit = None
+                elif (self._lower_limit is not None) and (lower_item < self._lower_limit):
+                    self._lower_limit = lower_item
+
+                if upper_item is None:
+                    self._upper_limit = None
+                elif (self._upper_limit is not None) and (upper_item > self._upper_limit):
+                    self._upper_limit = upper_item
+
+    @property
+    def precision(self):
+        return self._precision
+
+    @property
+    def scale(self):
+        return self._scale
+
+    def __repr__(self):
+        """
+        Human readable description of the range similar to a Python tuple.
+        """
+        if self.items:
+            result = "'%s'" % self
+        else:
+            result = str(None)
+        return result
+
+    def __str__(self):
+        """
+        Human readable description of the range similar to a Python tuple.
+        """
+        if self.items:
+            result = ""
+            is_first = True
+            for item in self._items:
+                if is_first:
+                    is_first = False
+                else:
+                    result += ", "
+                result += self._repr_item(item)
+        else:
+            result = str(None)
+        return result
+
+    def _repr_item(self, item):
+        """
+        Human readable description of a range item.
+        """
+        if item is not None:
+            result = ""
+            (lower, upper) = item
+            if lower is None:
+                assert upper is not None
+                result += ":%s" % upper.to_eng_string()
```

Comment:
To what end do we use `to_eng_string()` here? It can easily result in the decimal being displayed in the like '10E+6' which could be quite confusing to the end user.

A possible alternative might be string formats using `'%f'`, for example:

```python
>>> import decimal
>>> x = decimal.Decimal('123.45677')
>>> '%3.5f' % x
'123.45677'
```